# Airline Passenger Satisfaction Analysis

> **BY ABDULRAHEEM BASHIR**

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#import">Importing required files and libraries</a></li>
<li><a href="#invest">Data Inspection</a></li>
<li><a href="#clean">Data Cleaning</a></li>
<li><a href="#check">Sanity Check After Data Cleaning</a></li>
<li><a href="#explore">Data Exploration</a></li>
<li><a href="#rec">Recommendations</a></li>
</ul>

<a id='intro'></a>
## Introduction

> This project is the product of my continuous learning and getting my hands dirty with data analysis in Python.

**About Data**

This dataset was derived from the data playground of the Maven Analytics platform.

**Background info and context**

Customer satisfaction scores from 120,000+ airline passengers, including additional information about each passenger, their flight, and type of travel, as well as ther evaluation of different factors like cleanliness, comfort, service, and overall experience.

**Recommended Analysis:** The following are the questions I will be solving through out the analysis:
1. Which percentage of airline passengers are satisfied? Does it vary by customer type? What about type of travel?
2. What is the customer profile for a repeating airline passenger?
3. Does flight distance affect customer preferences or flight patterns?
4. Which factors contribute to customer satisfaction the most? What about dissatisfaction?

**Dataset:** The dataset contains one row for each passanger. The dataset can be downloaded from (Source: John D, via Kaggle).

**Attribute Information:** The following are the dataset descriptions:
- **ID:** Unique passenger identifier.
- **Gender:** Gender of the passenger (Female/Male).
- **Age:** Age of the passenger.
- **Customer Type:** Type of airline customer (First-time/Returning).
- **Type of Travel:** Purpose of the flight (Business/Personal).
- **Class:** Travel class in the airplane for the passenger seat.
- **Flight Distance:** Flight distance in miles.
- **Departure Delay:** Flight departure delay in minutes.
- **Arrival Delay:** Flight arrival delay in minutes.
- **Departure and Arrival Time Convenience:** Satisfaction level with the convenience of the flight departure and arrival times from 1 (lowest) to 5 (highest) - 0 means "not applicable".
- **Ease of Online Booking:** Satisfaction level with the online booking experience from 1 (lowest) to 5 (highest) - 0 means "not applicable".
- **Check-in Service:** Satisfaction level with the check-in service from 1 (lowest) to 5 (highest) - 0 means "not applicable".
- **Online Boarding:** Satisfaction level with the online boarding experience from 1 (lowest) to 5 (highest) - 0 means "not applicable".
- **Gate Location:** Satisfaction level with the gate location in the airport from 1 (lowest) to 5 (highest) - 0 means "not applicable".
- **On-board Service:** Satisfaction level with the on-boarding service in the airport from 1 (lowest) to 5 (highest) - 0 means "not applicable".
- **Seat Comfort:** Satisfaction level with the comfort of the airplane seat from 1 (lowest) to 5 (highest) - 0 means "not applicable".
- **Leg Room Service:** Satisfaction level with the leg room of the airplane seat from 1 (lowest) to 5 (highest) - 0 means "not applicable".
- **Cleanliness:** Satisfaction level with the cleanliness of the airplane from 1 (lowest) to 5 (highest) - 0 means "not applicable".
- **Food and Drink:** Satisfaction level with the food and drinks on the airplane from 1 (lowest) to 5 (highest) - 0 means "not applicable".
- **In-flight Service:** Satisfaction level with the in-flight service from 1 (lowest) to 5 (highest) - 0 means "not applicable".
- **In-flight Wifi Service:** Satisfaction level with the in-flight Wifi service from 1 (lowest) to 5 (highest) - 0 means "not applicable".
- **In-flight Entertainment:** Satisfaction level with the in-flight entertainment from 1 (lowest) to 5 (highest) - 0 means "not applicable".
- **Baggage Handling:** Satisfaction level with the baggage handling from the airline from 1 (lowest) to 5 (highest) - 0 means "not applicable".
- **Satisfaction:** Overall satisfaction level with the airline (Satisfied/Neutral or unsatisfied).

I will be providing insights base on the recommended analysis above.

<a id='import'></a>
## Importing files and libraries
> In this section, I will be Importing required files and libraries

In [2]:
# importing all necessary packages

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [3]:
# # Setting up the cells to display all Columns and rows

# pd.set_option('display.max_columns', None)
# pd.set_option('display.max_rows', None)

In [5]:
# reading in the data

data = pd.read_csv('airline_passenger_satisfaction.csv')

<a id='invest'></a>
## Data Inspection
> In this section, I will be scanning through the data so as to figure out the cleaning necessary on the data.

In [7]:
# displaying few top rows from the dataframe

data.head()

ID  Gender  Age Customer Type Type of Travel     Class  Flight Distance  \
0   1    Male   48    First-time       Business  Business              821   
1   2  Female   35     Returning       Business  Business              821   
2   3    Male   41     Returning       Business  Business              853   
3   4    Male   50     Returning       Business  Business             1905   
4   5  Female   49     Returning       Business  Business             3470   

   Departure Delay  Arrival Delay  Departure and Arrival Time Convenience  \
0                2            5.0                                       3   
1               26           39.0                                       2   
2                0            0.0                                       4   
3                0            0.0                                       2   
4                0            1.0                                       3   

   ...  On-board Service  Seat Comfort  Leg Room Service  Cleanliness  \
0  ...                 3             5                 2            5   
1  ...                 5             4                 5            5   
2  ...                 3             5                 3            5   
3  ...                 5             5                 5            4   
4  ...                 3             4                 4            5   

   Food and Drink  In-flight Service  In-flight Wifi Service  \
0               5                  5                       3   
1               3                  5                       2   
2               5                  3                       4   
3               4                  5                       2   
4               4                  3                       3   

   In-flight Entertainment  Baggage Handling             Satisfaction  
0                        5                 5  Neutral or Dissatisfied  
1                        5                 5                Satisfied  
2                        3                 3                Satisfied  
3                        5                 5                Satisfied  
4                        3                 3                Satisfied  

[5 rows x 24 columns]

In [8]:
# displaying few bottom rows from the dataframe

data.tail()

ID  Gender  Age Customer Type Type of Travel         Class  \
129875  129876    Male   28     Returning       Personal  Economy Plus   
129876  129877    Male   41     Returning       Personal  Economy Plus   
129877  129878    Male   42     Returning       Personal  Economy Plus   
129878  129879    Male   50     Returning       Personal  Economy Plus   
129879  129880  Female   20     Returning       Personal  Economy Plus   

        Flight Distance  Departure Delay  Arrival Delay  \
129875              447                2            3.0   
129876              308                0            0.0   
129877              337                6           14.0   
129878              337               31           22.0   
129879              337                0            0.0   

        Departure and Arrival Time Convenience  ...  On-board Service  \
129875                                       4  ...                 5   
129876                                       5  ...                 5   
129877                                       5  ...                 3   
129878                                       4  ...                 4   
129879                                       1  ...                 4   

        Seat Comfort  Leg Room Service  Cleanliness  Food and Drink  \
129875             1                 4            4               4   
129876             2                 5            2               2   
129877             3                 4            3               3   
129878             4                 5            3               3   
129879             2                 4            2               2   

        In-flight Service  In-flight Wifi Service  In-flight Entertainment  \
129875                  5                       4                        4   
129876                  4                       3                        2   
129877                  4                       2                        3   
129878                  4                       5                        3   
129879                  2                       3                        2   

        Baggage Handling             Satisfaction  
129875                 4  Neutral or Dissatisfied  
129876                 5  Neutral or Dissatisfied  
129877                 5  Neutral or Dissatisfied  
129878                 5                Satisfied  
129879                 1  Neutral or Dissatisfied  

[5 rows x 24 columns]

From displaying few top and bottom rows of the dataframe, I realized that:
- The column names have spaces between words; for easier referencing, I would want to use an underscore in place of those spaces.
- The majority of the columns use a 5-point rating scale to assess passenger satisfaction, which is represented by integers ranging from 1 to 5. I'd like to convert these to the string satisfaction form of a likert scale.

In [9]:
# displaying some information about the dataframe
# expecially to be sure that the datatype are accurate

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129880 entries, 0 to 129879
Data columns (total 24 columns):
 #   Column                                  Non-Null Count   Dtype  
---  ------                                  --------------   -----  
 0   ID                                      129880 non-null  int64  
 1   Gender                                  129880 non-null  object 
 2   Age                                     129880 non-null  int64  
 3   Customer Type                           129880 non-null  object 
 4   Type of Travel                          129880 non-null  object 
 5   Class                                   129880 non-null  object 
 6   Flight Distance                         129880 non-null  int64  
 7   Departure Delay                         129880 non-null  int64  
 8   Arrival Delay                           129487 non-null  float64
 9   Departure and Arrival Time Convenience  129880 non-null  int64  
 10  Ease of Online Booking                  1298

The output above shows that:
- The data comprise of satisfaction score form 129,880 passengers.
- Except for the 'Arrival Delay' column, which has 393 null entries, all columns have no null values.
- The 'Arrival Delay' column displays the flight arrival time in minutes. The datatype is float, but it should be integer.

In [20]:
data[data['Arrival Delay'].isnull()]

ID  Gender  Age Customer Type Type of Travel         Class  \
246        247    Male   11     Returning       Business      Business   
883        884    Male   39     Returning       Business      Business   
1965      1966    Male   36     Returning       Business       Economy   
2407      2408  Female   55     Returning       Business      Business   
2448      2449    Male   21     Returning       Personal  Economy Plus   
...        ...     ...  ...           ...            ...           ...   
129000  129001    Male   46    First-time       Business      Business   
129003  129004  Female   17    First-time       Business       Economy   
129014  129015    Male   57     Returning       Business      Business   
129015  129016  Female   22     Returning       Business      Business   
129488  129489    Male   23     Returning       Personal       Economy   

        Flight Distance  Departure Delay  Arrival Delay  \
246                 719               38            NaN   
883                 396                0            NaN   
1965                383                2            NaN   
2407               2904               58            NaN   
2448                767                5            NaN   
...                 ...              ...            ...   
129000             2704                7            NaN   
129003             2704                0            NaN   
129014             2704                0            NaN   
129015             2611                8            NaN   
129488             2704                0            NaN   

        Departure and Arrival Time Convenience  ...  On-board Service  \
246                                          1  ...                 4   
883                                          3  ...                 3   
1965                                         4  ...                 5   
2407                                         5  ...                 4   
2448                                         3  ...                 4   
...                                        ...  ...               ...   
129000                                       3  ...                 4   
129003                                       3  ...                 2   
129014                                       4  ...                 4   
129015                                       3  ...                 4   
129488                                       1  ...                 1   

        Seat Comfort  Leg Room Service  Cleanliness  Food and Drink  \
246                4                 3            4               4   
883                4                 3            3               3   
1965               2                 2            2               2   
2407               3                 4            4               5   
2448               1                 2            1               1   
...              ...               ...          ...             ...   
129000             4                 2            4               4   
129003             1                 4            1               1   
129014             4                 4            5               2   
129015             4                 3            4               4   
129488             2                 3            2               2   

        In-flight Service  In-flight Wifi Service  In-flight Entertainment  \
246                     3                       4                        4   
883                     3                       3                        3   
1965                    2                       2                        2   
2407                    4                       4                        4   
2448                    2                       1                        1   
...                   ...                     ...                      ...   
129000                  4                       3                        4   
129003                  3                       3           

The output above displays 393 rows with null values in the 'Arrival Delay' column. Null entries in the 'Arrival Delay' column will be replaced with 0 to avoid losing any data.

In [10]:
# Displaying summary statistics about the data

data.describe()

ID            Age  Flight Distance  Departure Delay  \
count  129880.000000  129880.000000    129880.000000    129880.000000   
mean    64940.500000      39.427957      1190.316392        14.713713   
std     37493.270818      15.119360       997.452477        38.071126   
min         1.000000       7.000000        31.000000         0.000000   
25%     32470.750000      27.000000       414.000000         0.000000   
50%     64940.500000      40.000000       844.000000         0.000000   
75%     97410.250000      51.000000      1744.000000        12.000000   
max    129880.000000      85.000000      4983.000000      1592.000000   

       Arrival Delay  Departure and Arrival Time Convenience  \
count  129487.000000                           129880.000000   
mean       15.091129                                3.057599   
std        38.465650                                1.526741   
min         0.000000                                0.000000   
25%         0.000000                                2.000000   
50%         0.000000                                3.000000   
75%        13.000000                                4.000000   
max      1584.000000                                5.000000   

       Ease of Online Booking  Check-in Service  Online Boarding  \
count           129880.000000     129880.000000    129880.000000   
mean                 2.756876          3.306267         3.252633   
std                  1.401740          1.266185         1.350719   
min                  0.000000          0.000000         0.000000   
25%                  2.000000          3.000000         2.000000   
50%                  3.000000          3.000000         3.000000   
75%                  4.000000          4.000000         4.000000   
max                  5.000000          5.000000         5.000000   

       Gate Location  On-board Service   Seat Comfort  Leg Room Service  \
count  129880.000000     129880.000000  129880.000000     129880.000000   
mean        2.976925          3.383023       3.441361          3.350878   
std         1.278520          1.287099       1.319289          1.316252   
min         0.000000          0.000000       0.000000          0.000000   
25%         2.000000          2.000000       2.000000          2.000000   
50%         3.000000          4.000000       4.000000          4.000000   
75%         4.000000          4.000000       5.000000          4.000000   
max         5.000000          5.000000       5.000000          5.000000   

         Cleanliness  Food and Drink  In-flight Service  \
count  129880.000000   129880.000000      129880.000000   
mean        3.286326        3.204774           3.642193   
std         1.313682        1.329933           1.176669   
min         0.000000        0.000000           0.000000   
25%         2.000000        2.000000           3.000000   
50%         3.000000        3.000000           4.000000   
75%         4.000000        4.000000           5.000000   
max         5.000000        5.000000           5.000000   

       In-flight Wifi Service  In-flight Entertainment  Baggage Handling  
count           129880.000000            129880.000000     129880.000000  
mean                 2.728696                 3.358077          3.632114  
std                  1.329340                 1.334049          1.180025  
min                  0.000000                 0.000000          1.000000  
25%                  2.000000                 2.000000          3.000000  
50%                  3.000000                 4.000000          4.000000  
75%                  4.000000                 4.000000          5.000000  
max                  5.000000                 5.000000          5.000000

Th summary statistic above reveals that:
- The passengers' minimum and maximum ages are 7 and 85 years, respectively. 25% of the passengers are under the age of 27, half of the passengers are under the age of 40, 75% are under the age of 51, and 25% are above the age of 51.
- The shortest and longest flying distances are 31 and 4,983 miles, respectively. Half of the flying distance is less than 844 miles, and the other half is greater.
- The minimum departure and arrival delay times are both 0 minutes, while the highest departure and arrival delay times are around 27 and 26 hours, respectively.
- 75% of departure and arrival delays are less than 12 minutes and 13 minutes, respectively.